_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,229 @@
+---
+author: hongbo
+date: "2020-06-10 18:00" 
+previewImg:
+category: compiler
+title: Overview of new encoding in BuckleScript
+description: |
+  Highlights of our newest changes to the internal representation of lazy values
+  and how it will benefit our users.
+---
+
+
+---
+title: An overview of new data representation in BuckleScript
+---
+
+In the next version of BuckleScript, we made several major changes to tweak the data representation for various data types, it's more idiomatic and debugger friendly. 
+
+Note since V8 or other JavaScript engines are tweaked to make idiomatic JS code running fast, so it also results in faster running code. 
+
+Another property, for a compiled language like BuckleScript, is that we can reason about [IC](https://en.wikipedia.org/wiki/Inline_caching) friendliness by just looking at the type definitions locally; this is very helpful for advanced users to write performance reliable JS code, if you are unfamiliar with how IC works in general, feel free to skip the section about IC.
+
+Note this article is quite dense so that we will skip the old encoding.
+
+## Record (stable)
+
+We compile records to idiomatic JS objects since [version 7](https://bucklescript.github.io/blog/2019/11/18/whats-new-in-7), this is great for performance and debugging. We also support label renaming to shorten field names to save sizes.
+
+Take the code below for example:
+
+```reasonml
+type int64 = {
+    loBits : int [@bs.as "lo"],
+    hiBits : int [@bs.as "hi]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+It will generate JS output as below:
+```js
+var value = {lo : 32, hi : 33}
+function rand (param){
+    return param.lo + param.hi
+}
+```
+
+If users want to make it even  shorter, they can choose to compile record as an array as below:
+
+```reasonml
+type int64 = {
+    loBits : int [@bs.as "0"],
+    hiBits : int [@bs.as "1"]
+}
+let value = {hiBits : 33 , loBits : 32 };
+let rand = ({loBits; hiBits}) => loBits + hiBits;
+```
+
+Now output JS as below:
+
+```js
+var value = [32,33]
+function rand(param){
+    return param[0] + param[1]
+}
+```
+
+The label renaming techniques can be applied systematically using a syntactic macro, in the future we may provide an advanced mode to apply it automatically. Another nice property is that only the type definition needs to be adapted, other parts of code is untouched.
+
+### IC friendliness
+
+Record is always in a perfect IC position, the compiler can ensure all generated records are of the same shape
```

Comment:
Records are always in a perfect position for Inline Caching (IC); the compiler can ensure all generated records are of the same shape.